lightning/src/routing/router.rs

```diff
@@ -861,19 +910,29 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 			match network.get_nodes().get(&pubkey) {
 				None => {},
 				Some(node) => {
-					add_entries_to_cheapest_to_target_node!(node, &pubkey, lowest_fee_to_node, value_contribution_msat);
+					add_entries_to_cheapest_to_target_node!(node, &pubkey, lowest_fee_to_node, value_contribution_msat, path_htlc_minimum_msat);
 				},
 			}
 		}
 
 		// Step (3).
-		// Stop either when recommended value is reached,
-		// or if during last iteration no new path was found.
-		// In the latter case, making another path finding attempt could not help,
-		// because we deterministically terminate the search due to low liquidity.
+		// Stop either when the recommended value is reached or if no new path was found in this
+		// iteration.
+		// In the latter case, making another path finding attempt won't help,
+		// because we deterministically terminated the search due to low liquidity.
 		if already_collected_value_msat >= recommended_value_msat || !found_new_path {
 			break 'paths_collection;
 		}
+		// Further, if this was our first walk of the graph, and we weren't limited by an
+		// htlc_minim_msat, return immediately because this path should suffice. If we were limited
+		// by an htlc_minimum_msat value, find another path with a higher value, potentially
+		// allowing us to pay fees to meet the htlc_minimum while still keeping a lower total fee.
+		if already_collected_value_msat == final_value_msat && payment_paths.len() == 1 {
+			if !hit_minimum_limit {
```

Comment:
Following tweak doesn't break router unit tests. It might be covered by new test `exact_fee_liquidity_limit()` but our overpaid value reduction logic is making the outcomes similar ?

```
diff --git a/lightning/src/routing/router.rs b/lightning/src/routing/router.rs
index 7f00ccdf..e555575b 100644
--- a/lightning/src/routing/router.rs
+++ b/lightning/src/routing/router.rs
@@ -928,9 +928,9 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
                // by an htlc_minimum_msat value, find another path with a higher value, potentially
                // allowing us to pay fees to meet the htlc_minimum while still keeping a lower total fee.
                if already_collected_value_msat == final_value_msat && payment_paths.len() == 1 {
-                       if !hit_minimum_limit {
-                               break 'paths_collection;
-                       }
+                       //if !hit_minimum_limit {
+                       //      break 'paths_collection;
+                       //}
                        path_value_msat = recommended_value_msat;
                }
        }
```

Reply:
I believe that line is a strict performance improvement - if we get there the path we found definitely should be the best one, we could do more work, but it won't find a better path.